File: 1000/LC 44.go
func isMatch(s string, p string) bool {
    m, n := len(s), len(p)
    f := make([][]bool, m + 1)
    for i := range f {
        f[i] = make([]bool, n + 1)
    }
    f[0][0] = true
    for i := 0; i <= m; i++ {
        for j := 1; j <= n; j++ {
            if p[j-1] == '*' {
                f[i][j] = f[i][j-1] || (i > 0 && f[i-1][j])
            } else {
                f[i][j] = i > 0 && f[i-1][j-1] && (s[i-1] == p[j-1] || p[j-1] == '?')
            }
        }
    }

    return f[m][n]
}

File: LC 44.cpp
class Solution {
public:
    bool isMatch(string s, string p) {
        int n = s.size(), m = p.size();
        vector<vector<int>>dp(n+1, vector<int>(m+1));
        for (int i = 1; i <= m; i++) {
            if (p[i-1] == '*')
                dp[0][i] = 1;
            else 
                break;
        }
        dp[0][0] = 1;
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= m; ++j) {
                if (p[j-1] == '?' || s[i-1] == p[j-1]) {
                    dp[i][j] = dp[i-1][j-1];
                } else if (p[j-1] == '*') {
                    dp[i][j] = dp[i-1][j] || dp[i][j-1];
                } 
            }
        }
        return dp[n][m];
    }
};